plugins/inputs/intel_pmu/activators.go

```diff
@@ -0,0 +1,205 @@
+//go:build linux && amd64
```

Comment:
The code in activators.go, config.go, resolver.go etc. is mostly manipulating intel_pmu plugin data structures built from config to run iaevents library methods. Accessing PMU counter data involves a lot of variables: event names, placement (core, uncore), grouping, aggregation possibilities that are all configurable by intel_pmu plugin. This is highly dependent on what the user of the plugin configures.

Additional interfaces on the top of the files are needed for proper mocking to achieve high test coverage. There is also a lot of nil-checking in code to be sure that every operation is safe. Most of the functionality is data manipulation, iteration over, or path choosing to pass plugin-specific user input to iaevents library methods.

Let's see what activation methods do we have in activators.go:

- activateEntities(...) -> iterate over all config segment provided by user
  -  activateCoreEvents(...) -> choose between perf group or not
     -  activateEventForPlacements(...) -> actually run a iaevents method to activate core events on specific placements (specific events on specific cores)
     -  activateCoreEventsGroup(...) ->  actually run a iaevents method to activate core events for specific perf group on specific placements (an event group is scheduled onto the CPU as a unit)
  -  activateUncoreEvents(....) -> iterate over events and over sockets to call activation method from iaevents for uncore type of events (many activated events for multi placements)

As you can see, the plugin doesn't make any lower level perf-related functionality, all of this is performed on iaevents side.

Reply:
Thank you for the explanation! I definitely don't have too much problem merging this PR as it is, so I don't want to seem too difficult but I am still a bit confused and it might be my misunderstanding of the scope of the package `iaevents`. Wouldn't a user using the `iaevents` package need to be able to configure the cores/events the same way as this plugin lets them? e.g. the `makeCorePlacements` method handles how the internal `ia.NewCorePlacements` should be called depending on the number of cores, wouldn't it make sense for `iaevents` to just provide the function `makeCorePlacements` ?

Reply:
Thanks for the suggestion - this actually may be a nice convenience method to be added into `iaevents` backlog for next release.

Of course the users of `iaevents` already have `NewCorePlacement(s)` methods to configure cores for events and how they use them depends on what users wants to accomplish - e.g. they may be interested only in using single core placement. The `makeCorePlacements` method is just a specific way of using `NewCorePlacement(s)` methods that align with what `intel_pmu` plugin wants to accomplish.

Also `makeCorePlacements` along with the `makeUncorePlacements` are designed to be easily mocked - which helps with good test coverage for the plugin

Reply:
Ok that makes sense, moving it as a convenience method in the future sound like a good idea but the logic living alongside the plugin should be fine for now because it does have good test coverage.